### PR TITLE
Initial model creation

### PR DESCRIPTION
**Pull request by antunmeg**
[Single-step Model(s)](https://app.gitkraken.com/glo/card/978546853c264c9ca04ffb8d1f40d1e7)

### CODE REVIEW

File: model.ipynb [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Separating features and labels\n",
+    "X = df.drop('PV', axis=1) # Features\n",
+    "y = df['PV'] # Labels"
```
Comment (sarajcev) [inline model.ipynb:172]: Ovo kako ste napravili matrice X i y nije analiza vremenskih serija (time-series model). Ovo je obična regresija (multiple regression). Treba prvo od vremenskih nizova podataka napraviti features (time-shift, diff, encode, etc.).

File: model.ipynb [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Performing train-test-split \n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)"
```
Comment (sarajcev) [inline model.ipynb:196]: Treba još razmisliti kako podijeliti podatke, s obzirom da je riječ o vremenskim serijama. Kako će random shuffle utjecati na vremenski niz podataka? Mislim da običan `train_test_split` neće ovdje biti dobar.

File: model.ipynb [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Using Grid Search for tuning Random Forest Regressor hyperparameters\n",
+    "time_start = timeit.default_timer()\n",
+    "parameters = {'n_estimators': [100, 500, 1000], 'max_depth': [1, 3, None]}\n",
+    "rfr_gs = GridSearchCV(estimator=RandomForestRegressor(criterion='mse'),\n",
```
Comment (sarajcev) [inline model.ipynb:208]: Možemo krenuti s RandomForest za početak.

**Review by sarajcev — CHANGES_REQUESTED** [review]

File: model.ipynb [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Resampling the dataset from 5-seconds to 15-minutes resolution (using mean)\n",
+    "pv = pv.resample('15min').mean()"
```
Comment (sarajcev) [inline model.ipynb:70]: Resample vremenske serije na rezoluciju od 15-min. Ako se bude radilo satno predviđanje (single-step ahead hourly forecast) mislim da bi bilo puno jednostavnije raditi sa podacima koji imaju satnu rezoluciju?

**Review by sarajcev — CHANGES_REQUESTED** [review]

**Review by sarajcev — CHANGES_REQUESTED** [review]
Napravio sam pregled ovoga što ste do sada napravili i stavio neke komentare kroz koje bi trebalo proći.

**Comment by sarajcev** [comment]
Evo samo za primjer funkcija koja generira features iz vremenskih serija. Mogu se dodati još neka druga obilježja s obzirom na karakteristike vremenskih serija. Uglavnom, na ovoj se bazi može dalje graditi.

```python
def engineer_features(dataframe, copy_data=True, resample=True):
    if copy_data:
        df = dataframe.copy()
    if resample:
        df = df.resample('1H').mean()
    
    # Engineer features from time-series data
    for col in df.columns:
        for i in range(1,25):
            # Shift data by lag of 1 to 24 hours
            df[col+'_{:d}h'.format(i)] = df[col].shift(periods=i)  # time-lag
        df[col+'_diff'] = df[col].diff()  # first-difference
    
    # Hour-of-day indicators with cyclical transform
    dayhour_ind = df.index.hour
    df['hr_sin'] = np.sin(dayhour_ind*(2.*np.pi/24))
    df['hr_cos'] = np.cos(dayhour_ind*(2.*np.pi/24))
    
    # Month indicators with cyclical transform
    month_ind = df.index.month
    df['mnth_sin'] = np.sin((month_ind-1)*(2.*np.pi/12))
    df['mnth_cos'] = np.cos((month_ind-1)*(2.*np.pi/12))
    
    # Encoding sunshine hours
    sun_ind = df['PV'] > 0.
    df['sun'] = sun_ind.astype(int)    

    # Drop rows with NaN values
    df.dropna(inplace=True)

    return df
```

**Comment by sarajcev** [comment]
**Pipeline**: SelectKbest + RandomForestRegressor uz RandomizedSearchCV (TimeSeriesSplit cross-validation) za **istovremenu** redukciju featuresa i optimizaciju hyper-parametara regresora:

```python
# Pipeline: SelectKBest and RandomForest
# SelectKBest is used for features reduction
selectbest = SelectKBest(score_func=mutual_info_regression, k='all')
forest = RandomForestRegressor(criterion='mse')
# Creating a pipeline
pipe = Pipeline(steps=[('kbest', selectbest), ('forest', forest)])
# Parameters of pipeline for the randomized search with cross-validation
param_dists = {'kbest__k': stats.randint(low=32, high=128), 
               'forest__n_estimators': stats.randint(low=200, high=1000),
               'forest__max_depth': [1, 3, 5, None]}
NITER = 100  # number of random search iterations
time_start = timeit.default_timer()
search = RandomizedSearchCV(estimator=pipe, param_distributions=param_dists, 
                            cv=TimeSeriesSplit(n_splits=3),
                            scoring='neg_mean_squared_error',
                            n_iter=NITER, refit=True, n_jobs=-1)
search.fit(X_train, y_train)
time_end = timeit.default_timer()
time_elapsed = time_end - time_start
print('Execution time (hour:min:sec): {}'.format(str(dt.timedelta(seconds=time_elapsed))))
print('Best parameter (CV score = {:.3f}):'.format(search.best_score_))
print(search.best_params_)
```

Ovo može služiti kao baza modela za daljnja razmatranja.

**Comment by sarajcev** [comment]
@antunmeg možeš pogledati primjere koda iz prethodna dva komentara (kao bazu za model). Nedostaje samo `train_test_split` koji treba implementirati tako da zadrži vremenski slijed podataka (bez random shuffle).

File: model.ipynb (hunk 1 above)
Comment (antunmeg) [reply model.ipynb:172]: Razumijem!

**Comment by antunmeg** [comment]
> Evo samo za primjer funkcija koja generira features iz vremenskih serija. Mogu se dodati još neka druga obilježja s obzirom na karakteristike vremenskih serija. Uglavnom, na ovoj se bazi može dalje graditi.
> 
> ```python
> def engineer_features(dataframe, copy_data=True, resample=True):
>     if copy_data:
>         df = dataframe.copy()
>     if resample:
>         df = df.resample('1H').mean()
>     
>     # Engineer features from time-series data
>     for col in df.columns:
>         for i in range(1,25):
>             # Shift data by lag of 1 to 24 hours
>             df[col+'_{:d}h'.format(i)] = df[col].shift(periods=i)  # time-lag
>         df[col+'_diff'] = df[col].diff()  # first-difference
>     
>     # Hour-of-day indicators with cyclical transform
>     dayhour_ind = df.index.hour
>     df['hr_sin'] = np.sin(dayhour_ind*(2.*np.pi/24))
>     df['hr_cos'] = np.cos(dayhour_ind*(2.*np.pi/24))
>     
>     # Month indicators with cyclical transform
>     month_ind = df.index.month
>     df['mnth_sin'] = np.sin((month_ind-1)*(2.*np.pi/12))
>     df['mnth_cos'] = np.cos((month_ind-1)*(2.*np.pi/12))
>     
>     # Encoding sunshine hours
>     sun_ind = df['PV'] > 0.
>     df['sun'] = sun_ind.astype(int)    
> 
>     # Drop rows with NaN values
>     df.dropna(inplace=True)
> 
>     return df
> ```
> 
> Ova funkcija vrati oko tristotinjak featuresa, koji se onda dalje mogu automatski reducirati ostavljajući najvažnije.

Shvatio sam logiku iza ovoga

File: model.ipynb (hunk 4 above)
Comment (antunmeg) [reply model.ipynb:70]: Slažem se

File: model.ipynb (hunk 2 above)
Comment (antunmeg) [reply model.ipynb:196]: Mogu za početak staviti parametar **shuffle=False** pa naknadno provjerimo kako utječe promjenom?

**Comment by antunmeg** [comment]
> **Pipeline**: SelectKbest + RandomForestRegressor uz RandomizedSearchCV (TimeSeriesSplit cross-validation) za **istovremenu** redukciju featuresa i optimizaciju hyper-parametara time-series regresora:
> 
> ```python
> # Pipeline: SelectKBest and RandomForest
> # SelectKBest is used for features reduction
> selectbest = SelectKBest(score_func=mutual_info_regression, k='all')
> forest = RandomForestRegressor(criterion='mse')
> # Creating a pipeline
> pipe = Pipeline(steps=[('kbest', selectbest), ('forest', forest)])
> # Parameters of pipeline for the randomized search with cross-validation
> param_dists = {'kbest__k': stats.randint(low=32, high=128), 
>                'forest__n_estimators': stats.randint(low=200, high=1000),
>                'forest__max_depth': [1, 3, 5, None]}
> NITER = 100  # number of random search iterations
> time_start = timeit.default_timer()
> search = RandomizedSearchCV(estimator=pipe, param_distributions=param_dists, 
>                             cv=TimeSeriesSplit(n_splits=3),
>                             scoring='neg_mean_squared_error',
>                             n_iter=NITER, refit=True, n_jobs=-1)
> search.fit(X_train, y_train)
> time_end = timeit.default_timer()
> time_elapsed = time_end - time_start
> print('Execution time (hour:min:sec): {}'.format(str(dt.timedelta(seconds=time_elapsed))))
> print('Best parameter (CV score = {:.3f}):'.format(search.best_score_))
> print(search.best_params_)
> ```
> 
> Ovo može služiti kao baza modela za daljnja razmatranja.

Ovo sam simulirao, rezultat je sljedeći:
Execution time (hour:min:sec): 0:13:20.189639
Best parameter (CV score = -0.001):
{'forest__max_depth': None, 'forest__n_estimators': 687, 'kbest__k': 33}

Jasno mi je da sada trebam kao parametre Random Forest Regressora staviti vrijednosti max_depth=None i n_estimators=687. Međutim, ne razumijem što točno znači ovaj broj 33. Ako imam 290 featuresa vjerojatno ovo govori da je odabranim kriterijem najbolje koristiti 33 featuresa, ali kojih 33? Treba mi informacija o nazivu featuresa da bi mogao izdvojiti najvažnije.

File: model.ipynb (hunk 2 above)
Comment (sarajcev) [reply model.ipynb]: Može.

File: model.ipynb [hunk 5]
```diff
@@ -202,30 +237,122 @@
...
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Choosing optimal hyperparameters from Grid Search output\n",
-    "rfr = RandomForestRegressor(criterion='mse', **rfr_gs.best_params_)\n",
+    "# Choosing optimal hyperparameters from RandomizedSearchCV output\n",
+    "rfr = RandomForestRegressor(criterion='mse', max_depth=None, n_estimators=687)\n",
```
Comment (sarajcev) [inline model.ipynb:315]: Ovo se ne radi ručno! `rfr = RandomForestRegressor(criterion='mse', **rfr_gs.best_params_)` Ovaj dio sa "**" služi kao tzv. argument unpacking. Sve treba biti automatizirano! Osim toga, korišten je argument `refit=True` kod `RandomizedSearchCV` što znači da `fit` metoda "vrača" najbolji regresor. Pogledajte malo detaljnije kako se koristi `scikit-learn` API. Konačno, `Pipeline` povezuje fazu selekcije najkorisnijih featuresa i treniranje regresora (featuresi će najvjerojatnije biti različiti kako se bude mjenjao vremenski horizont i cijeli sustav stoga mora biti potpuno automatiziran).

File: model.ipynb [hunk 6]
```diff
@@ -202,30 +237,122 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# # Using Grid Search for tuning Random Forest Regressor hyperparameters\n",
+    "# time_start = timeit.default_timer()\n",
+    "# parameters = {'n_estimators': [100, 500, 1000], 'max_depth': [1, 3, None]}\n",
+    "# rfr_gs = GridSearchCV(estimator=RandomForestRegressor(criterion='mse'),\n",
```
Comment (sarajcev) [inline model.ipynb:328]: `GridSearchCV` smo zamijenili s `RandomizedSearchCV` koji je učinkovitiji! Ovaj cijeli dio koda je redundantan!

**Review by sarajcev — CHANGES_REQUESTED** [review]
Kolega @antunmeg pogledajte malo detaljnije kako se koristi `scikit-learn` API, s posebnim naglaskom na `RandomizedSearchCV` i `Pipeline`s.

**Comment by sarajcev** [comment]
Dodao bih u **Pipeline** još jedan preprocessing korak:
```python
# Creating a pipeline
pipe = Pipeline(steps=[('preprocess', 'passthrough'), 
                       ('kbest', selectbest), 
                       ('forest', forest)])
# Parameters of pipeline for the randomized search with cross-validation
param_dists = {'preprocess': [None, StandardScaler()], 
               'kbest__k': stats.randint(low=32, high=128), 
               'forest__n_estimators': stats.randint(low=200, high=1000),
               'forest__max_depth': [1, 3, 5, None]}
```
da vidimo da li `StandardScaler` donosi poboljšanje (biti će automatski odabran ili pak neće).

File: model.ipynb [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.feature_selection import SelectKBest\n",
+    "from sklearn.feature_selection import mutual_info_regression\n",
+    "from sklearn.pipeline import Pipeline\n",
+    "from scipy import stats\n",
+    "from sklearn.model_selection import RandomizedSearchCV"
```
Comment (sarajcev) [inline model.ipynb:244]: Sve `import` naredbe moraju ići na vrh notebooka.

**Review by sarajcev — CHANGES_REQUESTED** [review]

**Comment by sarajcev** [comment]
@antunmeg Kako sada stvari stoje, model pretpostavlja da postoje dostupni podaci o vremenskoj prognozi za sat unaprijed kada se radi predikcija proizvodnje (hour-ahead single-step forecast). Kako mi radimo *ex-post* validaciju modela, to mi koristimo povijesne podatke i za ovo "predviđanje" vremenske prognoze. Ukoliko ne bi usvojili tu pretpostavku (o dostupnoj vremenskoj prognozi), tada bi trebalo nakon poziva `engineer_features` **eliminirati** sve varijable vezane uz vremensku prognozu `[CD, CM, CU, PREC, RH2m, SNOW, ST, SWD, SWDtop, TT2M, WS100m, WS10m]` i ostaviti samo njihove featurese (time-shift, diff). Time eliminiramo sve informacije o vremenkoj prognozi za sat (unaprijed) za koji radimo predikciju proizvodnje i ona se temelji isključivo na povijesnim podacima (o vremenskoj prognozi i proizvodnji). Model i dalje ostaje statičan.

File: model.ipynb [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+   "outputs": [],
+   "source": [
+    "# Visualize the predicted labels on test set\n",
+    "start = 1\n",
+    "end = start + 24*4\n",
+    "plt.figure(figsize=(15,8))\n",
+    "plt.plot(y_preds[start:end], ls='--', label='predictions')\n",
+    "plt.plot(y_test.iloc[start:end].values, label='actual')\n",
+    "plt.grid()\n",
+    "plt.legend();"
```
Comment (sarajcev) [inline model.ipynb:319]: Ovaj dio koda zamijeniti sa:
```python
plt.figure(figsize=(12,4.5))
plt.plot(y_test.index, y_preds, label='predictions')
plt.plot(y_test.index, y_test.values, lw=2, label='true values')
plt.legend()
plt.grid(axis='y')
plt.xlabel('Hours')
plt.show()
```

**Review by sarajcev — CHANGES_REQUESTED** [review]
